notebooks/utils/partially_upcast_nodes_to_fp32.py

```diff
@@ -0,0 +1,209 @@
+        orig.request = None
+        orig.compile()  # compile will set orig.request for OVModelForCausalLM
+        orig.request.start_async(inputs, shared_memory=shared_memory)
+
+    return new_start_async
+
+
+def partially_upcast_nodes_to_fp32(orig_model: Model, example_input: Union[List, Dict],
+                                   thresholds_per_op: Dict[str, Tuple] = None) -> Model:
+    model = orig_model.clone()  # todo: check if need to clone orig_models
+    nodes_to_track, outs_to_track, _ = insert_results_for_tracked_ops(model)
+    fp16_full_net_infer_values = infer_full_net_in_fp16(nodes_to_track, model, example_input)
+    fp16_infer_values = infer_nodes_in_fp16(nodes_to_track, fp16_full_net_infer_values)
+    fp32_infer_values = infer_nodes_in_fp32(nodes_to_track, fp16_full_net_infer_values)
+    del model
+    new_model = orig_model.clone()
+    mark_nodes_to_upcast_to_fp32(new_model, nodes_to_track, fp16_infer_values, fp32_infer_values, thresholds_per_op)
+    return new_model
+
+
+def insert_results_for_tracked_ops(model) -> (List, List, List):
+    # additional outputs to track inputs and output values of operations of interest
+    nodes_to_track = []
+    outputs = []
+    for i, op in enumerate(model.get_ordered_ops()):
+        if op.get_type_name() not in ops_to_track_map.keys():
+            continue
+        outputs.append(op.output(0))
+        nodes_to_track.append(op)
+
+        node_0 = op.input_value(0).get_node()
+        node_1 = op.input_value(1).get_node()
+        for node in [node_0, node_1]:
+            if node.get_type_name() != 'Constant' and not is_decompression_convert(node):  # for Consts we can take inputs from ov::Model
+                outputs.append(node.output(0))
+
+    orig_outputs = model.outputs.copy()
+    model.add_outputs(outputs)
+    return nodes_to_track, outputs, orig_outputs
+
+
+def get_const_value_from_ovmodel(node: Union[Constant, Node]) -> np.ndarray:
+    if node.get_type_name() == 'Constant':
+        assert node.get_element_type() == ov.Type.f32
+        return node.get_data()
+    elif is_decompression_convert(node):
+        # if model is compressed and constant values flow through decompression convert
+        const_node = node.input_value(0).get_node()
+        assert const_node.get_type_name() == 'Constant'
+        assert const_node.get_element_type().is_real()
+        return np.array(node.input_value(0).get_node().get_data(), dtype=np.float32)
+    else:
+        raise Exception(
+            f'Cannot get const values from ov.Model for node {node.get_friendly_name()} with type {node.get_type_name()}')
+
+
+def is_decompression_convert(node: Node) -> bool:
+    if node.get_type_name() != 'Convert':
+        return False
+    if len(node.get_rt_info()['is_decompression_0'].aslist()) > 0:
+        return True
+    return False
+
+
+def infer_full_net_in_fp16(nodes_to_track: List[Node], orig_model: ov.Model, example_inputs: List) -> List[Tuple]:
+    ie = Core()
```

Comment:
We have a suggestion not to use `ie` naming in notebooks. It should be 
`
import openvino as ov
core = ov.Core()
`

Reply:
agreed, and replaced `ie` to `core` everywhere in the file